Add missing env vars

harrietgrace opened the pull request:
Adds in the env vars that are in the job environment but aren't in the docs. Also added a section for deprecated vars. 

I've had a go at descriptions for these new ones, but some of them I couldn't find any info on so they're missing descriptions for now. If either of you could help me fill these in, that would be super! I've also had a guess at some default/examples for them. Let me know if any are inaccurate! 

- [x] Add All The Things
- [x] Figure out the overall page structure
- [x] Figure out how to explain hooks/modifications of envs
- [x] Check if we need to remove any after [buildkite-agent v3.9.0](https://github.com/buildkite/agent/releases/tag/v3.9.0) was released which filters out some config envs
- [x] Add .h4 styles
- [x] Add `BUILDKITE_AGENT_NAME` (was previously in the deprecated list, but not sure why?)
- [x] Add `BUILDKITE_AGENT_META_DATA_*` - mention `"BUILDKITE_AGENT_META_DATA_QUEUE=some-queue"` as that will be the most common appearing. It's set by buildkite-agent "tags" (which used to be called META_DATA)
- [x] Add `BUILDKITE_AGENT_EXPERIMENT=""`. It's blank if not set. It's `start --experiment`
- [x] Open a buildkite/agent issue to rename to `BUILDKITE_AGENT_TAG_` — and then we can update this doc to list `BUILDKITE_AGENT_META_DATA_*` as deprecated (https://github.com/buildkite/agent/issues/975)
- [x] Add issue for "Environment variable precedence", needs some reworking + split it off.
- [x] Fix all the ??? (`BUILDKITE_CLEAN_CHECKOUT`, `BUILDKITE_PLUGIN_VALIDATION`)
- [ ] Do a pass over all the envs to see if they need hook info and whether they can be modified

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -62,14 +98,42 @@ The following environment variables are automatically provided to every job, and
+  </tr>
+  <tr>
+    <th><code>BUILDKITE_HOOKS_PATH</code></th>
+    <td>A string of the path to the hooks directory on the agent machine. <p class="Docs__api-param-eg"><em>Default:</em> <code>"/buildkite/hooks"</code></p></td>
```
```suggestion
    <td>The path to the hooks directory. <p class="Docs__api-param-eg"><em>Example:</em> <code>"/etc/buildkite-agent/hooks/"</code></p></td>
```

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -78,6 +142,10 @@ The following environment variables are automatically provided to every job, and
   </tr>
+  <tr>
+    <th><code>BUILDKITE_LOCAL_HOOKS_ENABLED</code></th>
+    <td>A boolean to try and find local hooks for use instead of global hooks. <p class="Docs__api-param-eg"><em>Default:</em> <code>"false"</code></p></td>
```
```suggestion
    <td>Whether the agent is configured to allow running of local hooks. <p class="Docs__api-param-eg"><em>Values:</em> <code>"true"</code> and <code>"false"</code></p></td>
```

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -102,6 +170,18 @@ The following environment variables are automatically provided to every job, and
   </tr>
+  <tr>
+    <th><code>BUILDKITE_PLUGINS_ENABLED</code></th>
+    <td>A boolean to enable or disable the use of plugins when running a job on this agent. <p class="Docs__api-param-eg"><em>Default:</em> <code>"true"</code></p></td>
```
```suggestion
    <td>Whether the agent is configured to allow running of plugins. <p class="Docs__api-param-eg"><em>Values:</em> <code>"true"</code> and <code>"false"</code></p></td>
```

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -102,6 +170,18 @@ The following environment variables are automatically provided to every job, and
+  </tr>
+  <tr>
+    <th><code>BUILDKITE_PLUGIN_VALIDATION</code></th>
+    <td> <p class="Docs__api-param-eg"><em>Example:</em> <code>""</code></p></td>
```
```suggestion
    <td>Whether the agent is configured to validate plugin options. <p class="Docs__api-param-eg"><em>Values:</em> <code>"true"</code> and <code>"false"</code></p></td>
```

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -102,6 +170,18 @@ The following environment variables are automatically provided to every job, and
+  </tr>
+  <tr>
+    <th><code>BUILDKITE_PLUGINS_PATH</code></th>
+    <td>A string of the path to where the plugins will be checked out on the agent machine. <p class="Docs__api-param-eg"><em>Default:</em> <code>"/buildkite/plugins"</code></p></td>
```
```suggestion
    <td>The path to the plugins directory. <p class="Docs__api-param-eg"><em>Example:</em> <code>"/etc/buildkite-agent/plugins/"</code></p></td>
```

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -122,18 +202,30 @@ The following environment variables are automatically provided to every job, and
-    <th><code>BUILDKITE_REFSPEC</code></th>
-    <td>A custom refspec for the <code>buildkite-agent</code> bootstrap script to use when checking out code. <p class="Docs__api-param-eg"><em>Example:</em> <code>"+refs/weird/123abc:refs/local/weird/456"</code></p></td>
+    <th><code>BUILDKITE_RETRY_COUNT</code></th>
+    <td>An integer of how many times this job has been retried. <p class="Docs__api-param-eg"><em>Example:</em> <code>"0"</code></p></td>
```
Oh wow, this is one of the big standard ones that is missing! We should add this straight away perhaps?

```suggestion
    <td>How many times this job has been retried. <p class="Docs__api-param-eg"><em>Example:</em> <code>"0"</code></p></td>
```

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -146,6 +238,14 @@ The following environment variables are automatically provided to every job, and
     <td>The UUID of the build that <a href="/docs/pipelines/trigger-step">triggered</a> this build. <p class="Docs__api-param-eg"><em>Example:</em> <code>"5aa7c894-c8c0-435b-bc17-13923b90f163"</code>, or <code>""</code> if the build was not triggered from another build.</td>
   </tr>
+  <tr>
+    <th><code>BUILDKITE_TRIGGERED_FROM_BUILD_NUMBER</code></th>
```
Another big standard one that's missing!

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -146,6 +238,14 @@ The following environment variables are automatically provided to every job, and
+    <td>The number of the build that <a href="/docs/pipelines/trigger-step">triggered</a> this build. <p class="Docs__api-param-eg"><em>Example:</em> <code>"1264"</code>, or <code>""</code> if the build was not triggered from another build.</td>
+  </tr>
+  <tr>
+    <th><code>BUILDKITE_TRIGGERED_FROM_BUILD_PIPELINE_SLUG</code></th>
```
Another big standard one that's missing!

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -122,18 +202,30 @@ The following environment variables are automatically provided to every job, and
   </tr>
+  <tr>
+    <th><code>BUILDKITE_SSH_KEYSCAN</code></th>
+    <td>A boolean that describes whether or not to run `ssh-keyscan` before running `git checkout`. <p class="Docs__api-param-eg"><em>Default:</em> <code>"false"</code></p></td>
```
```suggestion
    <td>Whether the agent is configured to run `ssh-keyscan` before running `git checkout`. <p class="Docs__api-param-eg"><em>Values:</em> <code>"true"</code> and <code>"false"</code></p></td>
```

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -122,18 +202,30 @@ The following environment variables are automatically provided to every job, and
+  </tr>
+  <tr>
+    <th><code>BUILDKITE_SHELL</code></th>
+    <td>A string of the shell used to intepret build commands.  <p class="Docs__api-param-eg"><em>Default:</em> <code>"/bin/bash -e -c"</code></p></td>
```
```suggestion
   <td>The shell used to interpret build commands. <p class="Docs__api-param-eg"><em>Example:</em> <code>"/bin/bash -e -c"</code></p></td>
```

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -122,18 +202,30 @@ The following environment variables are automatically provided to every job, and
     <td>The UUID of the original build this was rebuilt from. <p class="Docs__api-param-eg"><em>Example:</em> <code>"1514"</code>, or <code>""</code> if not a rebuild.</p></td>
   </tr>
+  <tr>
+    <th><code>BUILDKITE_REFSPEC</code></th>
```
I think we already have this one in the list?

harrietgrace replied:
It was out of alphabetical order just slightly, I moved it into it's correct spot

toolmantim replied:
Ahhhh!

toolmantim replied:
Ahhhh!

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -165,6 +265,27 @@ The following environment variables are automatically provided to every job, and
+<tbody>
+  <tr>
+    <th><code>BUILDKITE_PROJECT_PROVIDER</code></th>
+    <td>A string of the source control provider. <p class="Docs__api-param-eg"><em>Example:</em> <code>"github"</code></p></td>
```
Perhaps we should just say this has been renamed?

```suggestion
    <td>This has been renamed to `BUILDKITE_PIPELINE_PROVIDER`.</td>
```

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -165,6 +265,27 @@ The following environment variables are automatically provided to every job, and
+  </tr>
+  <tr>
+    <th><code>BUILDKITE_PROJECT_SLUG</code></th>
+    <td>A string of the organization and pipeline names. <p class="Docs__api-param-eg"><em>Example:</em> <code>"acmeinc/build-and-test"</code></p></td>
```
```suggestion
    <td>This has been renamed to `BUILDKITE_PIPELINE_SLUG`.</td>
```

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -165,6 +265,27 @@ The following environment variables are automatically provided to every job, and
+  </tr>
+  <tr>
+    <th><code>BUILDKITE_SCRIPT_PATH</code></th>
+    <td>A string of the path to the script being run in this job. <p class="Docs__api-param-eg"><em>Example:</em> <code>"scripts/build.sh"</code></p></td>
```
```suggestion
    <td> This has been renamed to `BUILDKITE_COMMAND`</td>
```

harrietgrace commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -13,19 +13,55 @@ The following environment variables are automatically provided to every job, and
+  </tr>
+  <tr>
+    <th><code>BUILDKITE_AGENT_ID</code></th>
+    <td>A string of the ID of the agent that is running the job. <p class="Docs__api-param-eg"><em>Example:</em> <code>"1a222222-e999-3636-8ddd-802222222222"</code></p></td>
```
I think it's fine without it 👍🏻

harrietgrace commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -62,14 +98,42 @@ The following environment variables are automatically provided to every job, and
   </tr>
+  <tr>
+    <th><code>BUILDKITE_CONFIG_PATH</code></th>
+    <td> <p class="Docs__api-param-eg"><em>Default:</em> <code>"/buildkite/buildkite-agent.cfg"</code></p></td>
```
How come there's both BUILDKITE_AGENT_CONFIG and BUILDKITE_CONFIG_PATH?

harrietgrace replied:
I think this one got lost in all the comments 😅Are both of these vars actually used? Or is one from the job and one from the agent or something like that?

harrietgrace commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -62,14 +98,42 @@ The following environment variables are automatically provided to every job, and
   </tr>
+  <tr>
+    <th><code>BUILDKITE_ENV_FILE</code></th>
+    <td> <p class="Docs__api-param-eg"><em>Example:</em> <code>"/tmp/job-env-36711a2a-711a-484e-b180-e1b3711a80cf51b18711a"</code></p></td>
```
Do you want it in the list if it's internal? Is it something that people will see and wonder about?

toolmantim replied:
I just did some digging, and it's not actually internal at all… it was added to help people wanting to pass envs into Docker: https://github.com/buildkite/agent/pull/643 https://github.com/buildkite/agent/pull/682

So let's document it!

```suggestion
    <td>The path to the file containing the job’s environment variables. <p class="Docs__api-param-eg"><em>Example:</em> <code>"/tmp/job-env-36711a2a-711a-484e-b180-e1b3711a80cf51b18711a"</code></p></td>
```

We should probably document this feature, and link to it from this description though, because I don't think people will understand what this is from this description 🤔 At least it's a start I suppose, and they can go digging.

harrietgrace commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -62,14 +98,42 @@ The following environment variables are automatically provided to every job, and
+  </tr>
+  <tr>
+    <th><code>BUILDKITE_GIT_CLEAN_FLAGS</code></th>
+    <td>A string of the flags to use when running `git clean`. <p class="Docs__api-param-eg"><em>Default:</em> <code>"-ffxdq"</code></p></td>
```
I think that might be better for the plugins and hooks pages? List there the env vars that you can modify?

harrietgrace commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -13,19 +13,55 @@ The following environment variables are automatically provided to every job, and
+    <td>Name of the agent running the job. <p class="Docs__api-param-eg"><em>Example:</em> <code>"f3422e2537ca-1"</code></p></td>
+  </tr>
+  <tr>
+    <th><code>BUILDKITE_AGENT_TOKEN</code></th>
```
Will remove 👍🏻

harrietgrace commented:
What do you think about splitting up the main table into job vars and agent vars?

toolmantim commented:
Let's give splitting it a go?

lox commented:
Yeah, I reckon it would be worth splitting them into the environments that they are relevant. In the agent, there is env that is used for setting config parameters, then there is env that is exposed to a build from buildkite.com and then there is settings that can be modified with hooks or pipeline env to change the build behaviour. I reckon it absolutely merits differentiating them.

lox commented:
I'm deeply sorry, but I've written an essay on the subject: https://gist.github.com/lox/98f9b16b2ec6a7092f746556de50e72f

lox commented:
I've removed myself as a reviewer for now, but add me on the next iteration! Happy to help work through this.

harrietgrace commented:
I was thinking about the sections @lox had in his doc on env vars, and the sections that @toolmantim had mentioned in basecamp - what do you think about another section called something about env var subsets available in specific locations? It would contain tables with the vars available in pipeline.yml when you're creating it using buildkite.com, the vars you can use in the `buildkite-agent` commands, and any other special times when you can only access a subset of the full list. 

However, I think adding a section like that will take a while. I'd like to get the changes that are currently in this PR finished and merged first. Then we have a full list of env in our docs, and we can look at adding the new section in another PR and maybe even splitting out this doc into a reference doc and a more explanatory doc. 

Let me know what you think @toolmantim @lox! 😊

lox commented:
> what do you think about another section called something about env var subsets available in specific locations?

That was my preference, because it's the most accurate and clear on what is available where, but totally open to alternatives.

harrietgrace commented:
I've added a new section of 'bootstrap configuration variables', and finished off adding in the last few missing vars from the other sections. 

If all the descriptions are accurate and the examples are alright, I'd be keen to merge 🚀 In another PR I'll start the addition of the subsets section.

lox commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -165,6 +157,178 @@ The following environment variables are automatically provided to every job, and
+<table>
+<tbody>
+  <tr>
+    <th><code>BUILDKITE_NO_LOCAL_HOOKS</code></th>
```
Good question. The code is here:

https://github.com/buildkite/agent/blob/bc17c8b81079bfe2b942268142a34c3947144397/bootstrap/bootstrap.go#L312-L316

I might add a test and see.

lox replied:
Ah-hah. `BUILDKITE_NO_LOCAL_HOOKS=true` is respected, but `BUILDKITE_NO_LOCAL_HOOKS=false` does nothing.

lox replied:
That is confusing.

toolmantim replied:
Confusing, and logical… at the same time.

harrietgrace commented:
I had a look around at what other companies are doing for their env vars docs, and it's a pretty mixed bag. Some people do a big list in dot points, some people have tables, some people have the vars as headings and the explanations in the body.

I think that dot points work reasonably well, as long as they have nice padding around them ([cramped formatting](https://docs.travis-ci.com/user/environment-variables/#default-environment-variables) vs [spacious formatting](https://www.digitalocean.com/community/tutorials/how-to-read-and-set-environmental-and-shell-variables-on-a-linux-vps#common-environmental-and-shell-variables)).

I like how these two are split up into a [page full of headings](https://www.terraform.io/docs/configuration/environment-variables.html) with more in depth info/examples, and a [page with lots of words](https://www.terraform.io/docs/configuration/variables.html) about setting up and using the vars. 

However, I think that purely from a visual-parsing perspective, the [table format](https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables) is the easiest to use. I'm not sure how people are actually using our page though - are they coming here to scan for something they already know exists, or are they reading the descriptions and names to find out what they can do? The answer to this should dictate how we lay out the page.

If people are using the page as a reference, a nicely formatted list of dot points or a table would be great. If people are wanting info about each var and how to use it, having the vars as headings (and not having a toc at the top) would make more sense.

What do you think makes more sense for our audience @toolmantim @lox?

toolmantim commented:
Awesome research!

We might need to have a play with the different formats to see what feels right, but I think I'm leaning towards https://www.terraform.io/docs/configuration/environment-variables.html style environment variable names as subheadings, for what we need. Because similar to ours, the terraform ones can actually change behaviour and have subtle effects that you need to know about.

I'd also guess that if you're trying to achieve a task, you're only going to be using a few of the environment variables, and you'd probably prefer to know as much about them as possible, including multiple examples (which is important if you're writing regular expressions based on expected/possible values). So more details on each is probably better… even if half of them only have a one sentence description and one or two examples.

harrietgrace commented:
@toolmantim  I had a go at laying the variables out in the headings/body style, and it looks alright. I think maybe we'd need to figure out some different kind of styling for the variable headings to make the page a little more readable?

Here's what the top of the page looks like now, which is alright:
![image](https://user-images.githubusercontent.com/2074469/52577865-4fdb0e80-2e23-11e9-81d4-a5dcb68428e5.png)

✨🌱🇦🇺 (emojis to separate these very same-y sections) But then the different sections blur together a bit and it's a bit confusing further down:
![image](https://user-images.githubusercontent.com/2074469/52577881-5c5f6700-2e23-11e9-9c4e-cd33c0f58cfa.png)

toolmantim commented:
I think that works well as headers with a paragraph!

What did you think of removing the separate sections though, and just having one list? And then for each one covering the stuff I suggested earlier:

> * What's it called? (name) - done!
> * What is it? (description) - done!
> * Can I change it? (setting) And if so, where? Places you can change env vars include build env (what you'd provide in the "New Build" dialog), pipeline steps, hooks.

harrietgrace commented:
If we're going for just one big list, I'd like to add another bit in that says where that var comes from i.e. from the agent or the bootstrap or wherever it's set. I think that would be useful info, what do you think?

harrietgrace commented:
I had a go at merging some of @lox's info from his env vars post into each of the descriptions and I think it's worked reasonably well!

Here's an example of one of the vars that used to be in the Agent section:
![image](https://user-images.githubusercontent.com/2074469/52636520-2676bd00-2ecd-11e9-94aa-8d2e5b86a993.png)

I haven't put any extra bits in with the original 'job' variables, maybe that could say something about being able to set it in pipeline env etc.

lox commented:
I like that @harrietgrace, could we have standard thing of whether variables are "read only" or not and a link to what that means so we don't have to duplicate that info? 

I think I like "read only" vs "protected".

harrietgrace commented:
@toolmantim I got quite confused with "The value cannot be modified" in your example, it took me ages to figure out how you meant it 😅I thought it sounded like the value was set by Buildkite and can't be changed by the user at all (even after saying that it come from the agent config option - hence my confusion). 

I've spent like an hour trying to come up with alternatives, but all I've got so far is:
 "The value cannot be modified as an environment variable" - giving a little more context and making it clearer that it can be modified in the config file.
 "This variable is read only" - trying to explain that it's modifiable as a config option, but not when it's an env var.

I'm not happy with either of these, but I worry that other people will have the same confusion as I did with the simpler "the value cannot be modified".

toolmantim commented:
Yeah, I struggled with it too! “The value cannot be modified as an environment variable” is probably good enough? I figured the other agent configuration document would give them more confidence than this paragraph, even if there was doubt.

harrietgrace commented:
It looks like there's only a few different arrangements of content in the variable descriptions:

1. Has a description, can't be modified
![image](https://user-images.githubusercontent.com/2074469/52966235-110b0280-33a7-11e9-8d9e-ab600cb095f8.png)

1. Has a description, can be set in hooks
![image](https://user-images.githubusercontent.com/2074469/52966290-326bee80-33a7-11e9-8f08-2c9e52ce6016.png)

1. Links to a description, can't be modified
![image](https://user-images.githubusercontent.com/2074469/52966268-22540f00-33a7-11e9-98df-44a7a2e40545.png)

1. Is the opposite of the linked description, can't be modified
![image](https://user-images.githubusercontent.com/2074469/52966834-93e08d00-33a8-11e9-8d25-cd7e485115fe.png)

Questions:

Do we want to differentiate the ones that come from the bootstrap environment? Here's what I've got for those ones so far:
![image](https://user-images.githubusercontent.com/2074469/52966703-3ea47b80-33a8-11e9-9c1d-44052b44fe2e.png)

Do we want to mention anything about security in this secret key one? Also these ones have a bit about how you can set them in the pipeline env too. 
![image](https://user-images.githubusercontent.com/2074469/52967240-b7580780-33a9-11e9-8107-aee639d93636.png)

toolmantim commented:
I think the ability to go deep on each one is really great! They're looking ace.

The agent access token one perhaps should mention…

> The variable is read by the agent’s `artifact` and `meta-data` commands.

(you need to make sure you don't filter them out… say, with a Docker container, and still hope the commands work)

> Do we want to mention anything about security in this secret key one? Also these ones have a bit about how you can set them in the pipeline env too.

We definitely don't want to encourage people to be setting secrets in the pipeline env. I don't think we even want to mention it.

Maybe we don't need to differentiate the bootstrap ones with a generic sentence, but actually explain what the specific situation is with each one? For example, those two are both related to using private s3 buckets too, but we don't mention it.

What about something like:

> The secret access key for your S3 IAM user, for use with [private S3 buckets](https://buildkite.com/docs/agent/v3/cli-artifact#using-your-own-private-aws-s3-bucket). The variable is read by the `buildkite-agent artifact upload` command, and during the artifact upload phase of [command steps](https://buildkite.com/docs/pipelines/command-step#command-step-attributes). The access key must be set by exporting the environment variable in the `environment`, `pre-checkout` or `pre-command` hooks.

harrietgrace commented:
> Maybe we don't need to differentiate the bootstrap ones with a generic sentence, but actually explain what the specific situation is with each one?

That sounds cool! I'll definitely need some help with those bits though, everything I know about these vars is already in the doc 😅These are the ones I've listed as being "read from the bootstrap": https://gist.github.com/lox/98f9b16b2ec6a7092f746556de50e72f#bootstrap-run-time-configuration

@toolmantim I've added in your words for those two specific ones in your previous comment. @sj26 I've added your GitHub deployment vars in too 😊

toolmantim commented:
Happy to help get all those 9 bootstrappy options documented. If you wanna go and take a first stab, and then `@` @lox and I on which ones need looking at / documenting, we'll get those done. And then we might be close to 🚀?? 🎉🙌🏼

harrietgrace commented:
I've just noticed that there's a slew of 'opposite' style env vars that I haven't added, I'm guessing we want to include all these? We've already got BUILDKITE_NO_LOCAL_HOOKS.

BUILDKITE_NO_SSH_KEYSCAN
BUILDKITE_NO_COMMAND_EVAL
BUILDKITE_NO_PLUGINS
BUILDKITE_NO_PLUGIN_VALIDATION
BUILDKITE_NO_GIT_SUBMODULES,
BUILDKITE_NO_HTTP2

There's a bunch of other ones from the `buildkite-agent start` config section of @lox's document that I haven't added either 😬

toolmantim commented:
In the next Buildkite Agent release, none of the configuration environment variables will show up in builds anymore. Maybe we shouldn't include them in the doc now either?

harrietgrace commented:
The next point release or major release? If it's happening imminently, then totally happy to leave them out 👍🏻

lox commented:
It's in 3.9.0, which was just released @harrietgrace

lox commented:
Question: Why the backslashes before the underscores in the headers?

e.g `### BUILDKITE\_AGENT\_ACCESS_TOKEN`

lox commented:
Rather than "The value cannot be modified.", perhaps "Set by the agent"? Technically, it CAN be modified, it just doesn't do anything.

harrietgrace commented:
@lox It's parsed as markdown now that they're not in an html table, so if we don't escape the underscores it ends up partially italicised 😅They don't all need escaping, we just can't have two either side of a word.

toolmantim commented:
> Rather than "The value cannot be modified.", perhaps "Set by the agent"? Technically, it CAN be modified, it just doesn't do anything.

But won't any hook modification the env be reverted? I think for all intents and purposes it can't be modified in a way that gives it any affect.

lox commented:
> But won't any hook modification of the env be reverted/ignored? I think for all intents and purposes it can't be modified in a way that gives it any affect.

Yup, correct! The agent will reset this env every step and changing the value won't cause any agent config changes. I just find the statement "it can't be modified" confusing and ambiguous, but that could just be me. "Changes to this environment variable don't have any effect"? Up to you, it just didn't quite ring right to me, but I'm pretty deep in that world.

harrietgrace commented:
@lox I had trouble with the phrasing too, it took me ages of brainstorming other ideas to come around to @toolmantim's "the value cannot be modified". 

Another idea I had over the weekend was: "Changes will not be persisted" or something to that effect? Or even mention that we're the ones who set it, in a 'you can't set this' kind of way - "Variable set by Buildkite, changes will not be persisted".

harrietgrace commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -4,163 +4,448 @@ When the agent invokes your build scripts it passes in a set of standard Buildki
+
+### BUILDKITE\_CLEAN_CHECKOUT
+
+Whether the build should perform a clean checkout. This variable is read by ??? during ???. It can only be set by exporting the environment variable in the `environment` or `pre-checkout` hooks.
```
This one needs some help 😄

harrietgrace commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -4,163 +4,448 @@ When the agent invokes your build scripts it passes in a set of standard Buildki
+
+### BUILDKITE\_NO\_LOCAL_HOOKS
+
+Whether local hooks in the repository are allowed. This variable is read by the `buildkite-agent start` command ???. It can be set in the [agent config](https://buildkite.com/docs/agent/v3/configuration) file using the `no-local-hooks` option, or in by exporting this environment variable in the `environment`, `pre-checkout` or `pre-command` hooks.
```
Are we keeping NO_LOCAL_HOOKS? Or is that one of the ones that doesn't show up anymore?

harrietgrace commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -4,163 +4,448 @@ When the agent invokes your build scripts it passes in a set of standard Buildki
+
+### BUILDKITE\_PLUGIN_VALIDATION
+
+Whether the agent is configured to validate plugin options. This variable is generated by the `buildkite-agent start` command for use by the `buildkite-agent bootstrap` command ???
```
This one needs some help too

toolmantim commented:
Let's video pair on this one! 🎉

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -4,163 +4,448 @@ When the agent invokes your build scripts it passes in a set of standard Buildki
+
+### BUILDKITE\_GITHUB\_DEPLOYMENT_ID
+
+The GitHub deployment ID. Only available on builds triggered by a [GitHub Deployment](https://developer.github.com/v3/repos/deployments/).
```
Can this be modified in hooks?

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -4,163 +4,448 @@ When the agent invokes your build scripts it passes in a set of standard Buildki
+
+### BUILDKITE\_GITHUB\_DEPLOYMENT_ENVIRONMENT
+
+The name of the GitHub deployment environment. Only available on builds triggered by a [GitHub Deployment](https://developer.github.com/v3/repos/deployments/).
```
Can this be modified in hooks?

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -4,163 +4,448 @@ When the agent invokes your build scripts it passes in a set of standard Buildki
+
+### BUILDKITE\_GITHUB\_DEPLOYMENT_TASK
+
+The name of the GitHub deployment task. Only available on builds triggered by a [GitHub Deployment](https://developer.github.com/v3/repos/deployments/).
```
Can this be modified in hooks?

toolmantim commented on pages/pipelines/environment_variables.md.erb:
```diff
@@ -4,163 +4,448 @@ When the agent invokes your build scripts it passes in a set of standard Buildki
+
+### BUILDKITE\_GITHUB\_DEPLOYMENT_PAYLOAD
+
+The GitHub deployment payload data as serialized JSON. Only available on builds triggered by a [GitHub Deployment](https://developer.github.com/v3/repos/deployments/).
```
Can this be modified in hooks?
